# tpmi
## 1. Phenotyping
### 1.1 Diseases and symptoms (phecode: https://phewascatalog.org/)
### 1.2 Anthropometric and laboratory measurements
## 2. Genotyping QC and imputation 
## 3. Genome-wide association study
### 3.1 PCAir, PCRelate, and PRIMUS (GENESIS: https://bioconductor.org/packages/release/bioc/html/GENESIS.html; PRIMUS: https://primus.gs.washington.edu/primusweb/)
### 3.2 Generalized linear mixed model (SAIGE: https://saigegit.github.io/SAIGE-doc/)
### 3.3 Generalized linear model (PLINK2: https://www.cog-genomics.org/plink/2.0/)
## 4. Post GWAS analysis
### 4.1 Known loci replication (PGRM: https://github.com/PheWAS/pgrm)
### 4.2 Fine-mapping (SuSie: https://stephenslab.github.io/susieR/index.html) 
### 4.3 Heritability estimation (LDSC: https://github.com/bulik/ldsc)
### 4.4 Gene-level heritability estimation (h2gene)
### 4.4 Colocalization (with eQTL and pairwise, coloc: https://chr1swallace.github.io/coloc/index.html, GTeX v.8)
### 4.5 Genetic correlation and Clustering (LDSC: phenome-wide; popcorn: across populations https://github.com/brielin/Popcorn)
## 5. PRS
### 5.1 Individual phenotype (LDPred2, Lassosum2, PRScs, MegaPRS, SBayesR)
### 5.2 Mult-phenotype (PRSmix: https://github.com/buutrg/PRSmix)
### 5.3 Performance evaluation (R^2 and AUC)
